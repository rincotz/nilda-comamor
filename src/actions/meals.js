import db, { storage, auth, firebase } from '../firebase/firebase'
import {ADD_MEAL, EDIT_MEAL, REMOVE_MEAL, SET_MEALS} from "./constants";

// ADD_MEAL
export const addMeal = (meal) => ({
    type: ADD_MEAL,
    meal
})

export const startAddMeal = (mealData = {}) => {
    return (dispatch) => {
        const {
            name,
            description,
            price,
            available,
            location,
            chef,
            courrier,
            courrierStart,
            courrierEnd,
            pickUp,
            pickUpStart,
            pickUpEnd,
            table,
            tableStart,
            tableEnd,
            frozen,
            picture
        } = mealData
        const meal = {
            name,
            description,
            price,
            available,
            location,
            chef,
            courrier,
            courrierStart,
            courrierEnd,
            pickUp,
            pickUpStart,
            pickUpEnd,
            table,
            tableStart,
            tableEnd,
            frozen,
            picture
        }

        return db.collection('meals')
            .add(meal)
            .then((docRef) => {
                dispatch(addMeal({
                    id: docRef.id,
                    ...meal
                }))
            })
    }
}

export const getCooker = () => {
    auth.onAuthStateChanged(cooker => {
        return db.collection('users').doc(cooker.uid)
    })
}

export const addMealPic = (uid, blob, mealImage) => {
    const storageRef = firebase.storage().ref()
    storageRef.child(`meals/${uid}`).put(blob, { contentType: mealImage.type }).then(snapshot => {
        return snapshot.ref.getDownloadURL()
    })
}

// EDIT_MEAL
export const editMeal = (id, updates) => ({
    type: EDIT_MEAL,
    id,
    updates
})

export const startEditMeal = (id, updates) => {
    return (dispatch) => {
        return db.collection('meals').doc(id).update(updates)
            .then(() => {
                dispatch(editMeal(id, updates))
            })
    }
}

// REMOVE_MEAL
export const removeMeal = ({ id }) => ({
    type: REMOVE_MEAL,
    id
})

export const startRemoveMeal = ({ id }) => {
    return (dispatch) => {
        return db.collection('meals').doc(id).delete()
            .then(dispatch(removeMeal({ id })))
    }
}

// SET_MEALS
export const setMeals = (meals) => ({
    type: SET_MEALS,
    meals
})

export const startSetMeals = () => {
    return (dispatch) => {
        return db.collection('meals')
            .get()
            .then((querySnapshot) => {
                const meals = []
                querySnapshot.forEach((doc) => {
                    meals.push({
                        id: doc.id,
                        ...doc.data()
                    })
                })
                dispatch(setMeals(meals))
            })
    }
}