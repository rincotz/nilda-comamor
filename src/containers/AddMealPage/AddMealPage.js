import React from 'react';
import { connect } from 'react-redux'
import FormLabel from '@material-ui/core/FormLabel'
import MealForm from "../MealForm/MealForm";
import { startAddMeal } from "../../actions/meals";
import selectMeals from '../../selectors/selectMeals'

export class AddMealPage extends React.Component {
    onSubmit = (meal) => {
        this.props.startAddMeal(meal)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <FormLabel component='legend'>Cadastre seu prato:</FormLabel>
                <MealForm
                    onSubmit={this.onSubmit}
                    partner={this.props.partner}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    partner: state.partner
})

const mapDispatchToProps = (dispatch) => ({
    startAddMeal: (meal) => dispatch(startAddMeal(meal))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddMealPage)
