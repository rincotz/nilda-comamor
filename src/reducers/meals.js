import {ADD_MEAL, EDIT_MEAL, REMOVE_MEAL, SET_MEALS} from "../actions/constants";

// Meals Reducer
export const mealDefaultState = {
    id: "",
    available: "",
    chef: {
        uid: "",
        id: "",
        name: "",
        gender: "",
        birth: "",
        email: "",
        phone: "",
        bio: "",
        nationality: "",
        addressLine1: "",
        addressLine2: "",
        number: "",
        neighborhood: "",
        zip: "",
        picture: "",
        geoPoint: "",
        guestRatings: "",
        cookerRatings: "",
        ratings: ""
    },
    courrier: "",
    courrierEnd: "",
    courrierStart: "",
    description: "",
    frozen: "",
    name: "",
    pickUp: "",
    pickUpEnd: "",
    pickUpStart: "",
    picture: "",
    price: "",
    table: "",
    tableEnd: "",
    tableStart: ""
}
const mealsReducerDefaultState = []
export default (state = mealsReducerDefaultState, action) => {
    switch (action.type) {
        case ADD_MEAL:
            return [...state, action.meal]
        case REMOVE_MEAL:
            return state.filter(({ id }) => (
                id !== action.id
            ))
        case EDIT_MEAL:
            return state.map((meal) => {
                if (meal.id === action.id) {
                    return {
                        ...meal,
                        ...action.updates
                    }
                } else {
                    return meal
                }
            })
        case SET_MEALS:
            return action.meals
        default:
            return state
    }
}