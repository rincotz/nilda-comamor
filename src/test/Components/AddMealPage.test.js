import React from 'react'
import { shallow } from 'enzyme/build'
import { AddMealPage } from "../../containers/AddMealPage/AddMealPage";
import meals from '../fixtures/meals'

let startAddMeal, history, wrapper

beforeEach(() => {
    startAddMeal = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<AddMealPage startAddMeal={startAddMeal} history={history} />)
})

test('should render AddPartnerPage correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle onSubmit', () => {
    wrapper.find('MealForm').prop('onSubmit')(meals[1])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(startAddMeal).toHaveBeenLastCalledWith(meals[1])
})