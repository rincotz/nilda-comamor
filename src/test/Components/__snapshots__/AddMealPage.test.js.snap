// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddPartnerPage correctly 1`] = `
<div>
  <h1>
    Cadastre seu prato
  </h1>
  <MealForm
    onSubmit={[Function]}
  />
</div>
`;
